src/server.rs

```diff
@@ -88,7 +104,50 @@ async fn websocket_handler(
     ws: WebSocketUpgrade,
     ConnectInfo(addr): ConnectInfo<SocketAddr>,
     headers: HeaderMap,
-) -> impl IntoResponse {
+) -> Response {
+    // If API keys are required but none provided in URL path, reject
+    if !state.api_keys.is_empty() {
+        state.metrics.unauthorized_requests.increment(1);
+        return Response::builder()
+            .status(StatusCode::UNAUTHORIZED)
+            .body(Body::from(
+                json!({"message": "API key required"}).to_string(),
+            ))
+            .unwrap();
+    }
```

Comment:
Instead of having this check, why not just disable the unauthenticated route completely?